From 6d8c1d5e86093530ccf0870264f6fc0a1fe17366 Mon Sep 17 00:00:00 2001
From: Ang Tien Sung <tien.sung.ang@intel.com>
Date: Wed, 1 Dec 2021 15:37:20 +0800
Subject: [PATCH 18/20] HSD #14015013554: firmware: stratix10-svc: Fixed thread
 stop

commit 8f925167e377c42c71c3efc43aac4c932691826e from
https://github.com/altera-opensource/linux-socfpga/commits/socfpga-5.4.124-lts

Ensures that the thread task is checked for validity before calling kthread_stop
when the user aborts.

Signed-off-by: Ang Tien Sung <tien.sung.ang@intel.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/firmware/stratix10-svc.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/firmware/stratix10-svc.c b/drivers/firmware/stratix10-svc.c
index a7f5d12aff7f..85a8ec247a43 100644
--- a/drivers/firmware/stratix10-svc.c
+++ b/drivers/firmware/stratix10-svc.c
@@ -1436,11 +1436,20 @@ void stratix10_svc_done(struct stratix10_svc_chan *chan)
 {
 	/* stop thread when thread is running */
 	if (chan->task) {
-		pr_debug("%s: %s: svc_smc_hvc_shm_thread is stopped\n",
-		__func__, chan->name);
-		kthread_stop(chan->task);
+
+		if (!IS_ERR(chan->task)) {
+			struct task_struct *task_to_stop = chan->task;
+
+			chan->task = NULL;
+			pr_debug("%s: %s: svc_smc_hvc_shm_thread is stopping\n",
+					__func__, chan->name);
+			kthread_stop(task_to_stop);
+		}
+
 		chan->task = NULL;
 	}
+	pr_debug("%s: %s: svc_smc_hvc_shm_thread has stopped\n",
+					__func__, chan->name);
 }
 EXPORT_SYMBOL_GPL(stratix10_svc_done);
 
-- 
2.31.1

